### PR TITLE
Okina memory backends: Std, Cuda, Unified, Umpire and Protected [okina-mmu]

### PR DESCRIPTION
**Pull request by camierjs**
Add different memory back-ends and 'features': CUDA, UVM, Umpire, protected and aligned.
Add a `debug` device that uses the code path of the devices, but stays on the host.
The `protected` mode allowes to get an error when `host` side pointers are used while the data is on the `device` side.

- [x] Update CHANGELOG to mention
   - Umpire and UVM support
   - Debug device + PROTECTED memory space
   - New device options and priority
- [x] the `debug` device should be exclusive
- [x] delete fix in PROTECTED mode

### CODE REVIEW

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -5828,7 +5828,6 @@ void Mesh::SetNodes(const Vector &node_coord)
 void Mesh::NewNodes(GridFunction &nodes, bool make_owner)
 {
    if (own_nodes) { delete Nodes; }
-   nodes.Pull();
```
Comment (tzanio) [inline mesh/mesh.cpp]: Why is this no longer necessary?

Comment (camierjs) [reply mesh/mesh.cpp]: I have moved it to for now on the application's side, like in [Laghos](https://github.com/CEED/Laghos/blob/3d35ede905bd5f8259f3f500c00620ce255b2212/laghos_solver.cpp#L411).
I wanted to reduce the number of `Push`/`Pull`, to get a better understanding where they are really needed. We can re-introduce the ones that really make sense.

File: general/device.cpp [hunk 2]
```diff
@@ -24,21 +24,21 @@ namespace mfem
...
 };
 
 // Backend names listed by priority, high to low:
 static const char *backend_name[Backend::NUM_BACKENDS] =
 {
-   "occa-cuda", "raja-cuda", "cuda", "occa-omp", "raja-omp", "omp",
-   "occa-cpu", "raja-cpu", "cpu"
+   "occa-cuda", "raja-cuda", "cuda", "cuda-uvm",
+   "occa-omp", "raja-omp", "omp",
+   "occa-cpu", "raja-cpu", "debug", "cpu"
```
Comment (tzanio) [inline general/device.cpp]: Is it still true that this is the order of priority?

Comment (camierjs) [reply general/device.cpp]: the `debug` device should be exclusive: I need to add the test, you are right.
The priority should be fine then.

File: general/mem_manager.cpp [hunk 3]
```diff
@@ -11,11 +11,22 @@
...
 #include <list>
+#include <cstring> // std::memcpy
 #include <unordered_map>
 
+#include <signal.h>
+
+#ifndef _WIN32
+#include <sys/mman.h>
+#else
+#define posix_memalign(p,a,s) (((*(p))=_aligned_malloc((s),(a))),*(p)?0:errno)
```
Comment (tzanio) [inline general/mem_manager.cpp]: Has this been tested on Windows?

File: general/mem_manager.cpp [hunk 4]
```diff
@@ -26,26 +37,37 @@ namespace internal
 struct Alias;
 
 /// Memory class that holds:
-///   - a boolean telling which memory space is being used
 ///   - the size in bytes of this memory region,
 ///   - the host and the device pointer,
-///   - a list of all aliases seen using this region (used only to free them).
+///   - a list of all aliases seen using this region (used only to free them),
+///   - a boolean telling which memory space is being used,
+///   - another boolean to indicate if unified memory is used.
```
Comment (tzanio) [inline general/mem_manager.cpp]: Could be better to move these comments to the class members.

File: general/mem_manager.cpp (hunk 3 above)
Comment (camierjs) [reply general/mem_manager.cpp]: Yes, this is meant for `WIN32`.

File: general/mem_manager.cpp [hunk 5]
```diff
@@ -61,17 +83,424 @@ struct Ledger
...
 static internal::Ledger *maps;
 
+namespace internal
+{
+
+static bool managed;
+
+// *****************************************************************************
+/// The memory space abstract class: one pair (host + device) memory space
+class MemorySpace
```
Comment (tzanio) [inline general/mem_manager.cpp]: Can we pick a different name? It is confusing to have a class and a structure with the same name

File: general/mem_manager.cpp [hunk 6]
```diff
@@ -61,17 +83,424 @@ struct Ledger
...
+   { rm.copy(dst, const_cast<void*>(src), bytes); return dst; }
+};
+#endif // MFEM_USE_UMPIRE 
+#endif // MFEM_USE_CUDA
+
+// *****************************************************************************
+
+/// (STD + NONE) memory space
+class StdNoneMemorySpace : public StdHostMemorySpace,
+   public NoneDeviceMemorySpace {};
```
Comment (tzanio) [inline general/mem_manager.cpp]: Defining a new class for each pair of Host and Device memory spaces seems like an overkill. Can't we just have one class with two variables inside?

**Comment by tzanio** [comment]
Update `CHANGELOG` to mention
- Umpire and UVM support
- Debug device + PROTECTED memory space
- New device options and priority

**Comment by tzanio** [comment]
Approved modulus TODO items.

Thanks for adding these capabilities @camierjs!

File: general/device.cpp [hunk 7]
```diff
@@ -24,21 +24,21 @@ namespace mfem
 namespace internal
 {
 
-OccaDevice occaDevice;
-
 // Backends listed by priority, high to low:
 static const Backend::Id backend_list[Backend::NUM_BACKENDS] =
 {
-   Backend::OCCA_CUDA, Backend::RAJA_CUDA, Backend::CUDA,
+   Backend::OCCA_CUDA, Backend::RAJA_CUDA, Backend::CUDA, Backend::CUDA_UVM,
```
Comment (acfisher) [inline general/device.cpp]: It's good to see a way to support UVM coming in.  If worse comes to worst we could go for HYPRE GPU support with this.

Comment (acfisher) [reply general/device.cpp]: It's not in any of these changes, but I took a look at the code for the Push and Pull when using UVM.  I noticed that you are calling cudaMemcpy for those which could cause trouble when one of the memory spaces is oversubscribed.  I would recommend switching these to cudaMemAdvise.

File: general/mem_manager.cpp (hunk 5 above)
Comment (camierjs) [reply general/mem_manager.cpp]: I changed that to `Memory::Type`, but I'm open to other suggestions!

File: general/mem_manager.cpp (hunk 6 above)
Comment (camierjs) [reply general/mem_manager.cpp]: I updated the code with your approach: it's better that way, thanks!

File: general/device.cpp (hunk 2 above)
Comment (camierjs) [reply general/device.cpp]: Moved to the todo list

**Comment by artv3** [comment]
Hows this PR coming along? If we have Umpire doing the memory management here, we can have our different packages talk to each other via umpire. Very excited! 👍

**Comment by acfisher** [comment]
We are working to pull it in for our imminent release.  Stay tuned!

**Pull request closed without merging**